docs/content/sips/011-component-versioning.md

```diff
@@ -0,0 +1,41 @@
+
+Note that when we say "SDK" in this document, we mean either one of the official Spin SDKs _or_ raw bindings generated from the `WIT` files and the `wit-bindgen` version shipped with a given version of Spin, e.g. for languages which do not yet have an "official" SDK.
+
+## Proposal
+
+Although `WIT` does not yet have a way to express API versions, we can emulate it by introducing a new interface called `spin-sdk-version.wit` containing a single function to be exported by the guest component:
+
+```
+record version-details {
+  commmit-hash: string,
+  sdk-language: string
+}
+
+spin-sdk-version-$MAJOR-$MINOR: func() -> version-details
+```
+
+where `$MAJOR` and `$MINOR` are the Spin major and minor version numbers the SDK targets. The intention behind embedding these numbers in the name of the function is that it can be checked statically without instantiating or running the component, and it is immune to ABI changes.  The `version-details` record provides additional details about the SDK used to build the component.  Getting that information requires actually running the component, which may not be possible if the host does not support the ABI for that version, so it won't be useful in all scenarios.  (TODO: should we just encode everything in the name?)
+
+When Spin deploys or runs an application, it will have an opportunity to check the versions of its component(s) for compatibility, resulting in one of the following scenarios:
+
+- *No version found*: accept and run the component, raising an error if an incompatibility is detected.  This matches the current behavior.
+
+- *Version found, and it's within the range that this build of Spin supports*: accept and run the component.  Presumably no ABI or API incompatibilities are possible in this scenario, unless the component is reporting an inaccurate version or a breaking SDK change was made without bumping the version.
+
+- *Version found, and it's _not_ within the range that this build of Spin supports*: by default, optimistically accept and run the component, only raising an error if and when an ABI or API incompatibility is detected.  The error message should contain a clear, actionable report based on the component's version.  Alternatively, if "strict" mode is requested (e.g. via a command line option), Spin should reject the application immediately without attempting to deploy or run it, again providing a clear error message.
```

Comment:
Oh if we are going to execute by default anyway then maybe my concern above goes away.

Reply:
Yeah I imagine this being primarily about nice error messages and perhaps compatibility hacks in the near term.